lib/ahoy_activity_utils.rb

```diff
@@ -1,83 +1,91 @@
-        puts "\tUser ID: #{user}"
+      # We group by collection first, so we can sort and sum the timestamps
+      events.group_by { |e| e.properties['collection_id'] }.each do |cid, event|
+        timestamps = event.map { |e| [e[:time], e[:name]] }
 
-        # We figure if you are logged in and doing stuff it counts as volunteer time
-        # This could be expanded or resctricted
-        events = Ahoy::Event
-            .where(user_id: user)
-            .where("time BETWEEN ? AND ?", day.beginning_of_day, day.end_of_day)
-            .select(:id, :user_id, :time, :name, :properties)
+        minutes = total_contiguous_seconds(timestamps) / 60
 
-        unless events.empty?
-            # We group by collection first, so we can sort and sum the timestamps
-            events
-                .group_by { |e| e.properties["collection_id"] }
-                .each do |cid, event|
-                    timestamps = event.map{|e| [e[:time], e[:name]] }
+        next unless minutes.positive?
 
-                    minutes = self.total_contiguous_seconds(timestamps) / 60
-                    
-                    unless minutes <= 0
-                        begin 
-                            activity = AhoyActivitySummary.create({
-                                date: day.beginning_of_day,
-                                user_id: user,
-                                collection_id: cid,
-                                activity: 'transcribe',
-                                minutes: minutes
-                            })
-                        rescue ActiveRecord::RecordNotUnique => e
-                            puts e
-                        else
-                            puts "\t\t- Collection: #{activity.collection_id} \t#{activity.minutes} minutes\n"
-                        end
-                    end
-                end
-            end
+        begin
+          activity = AhoyActivitySummary.create(
+            {
+              date: day.beginning_of_day,
+              user_id: user,
+              collection_id: cid,
+              activity: 'transcribe',
+              minutes: minutes
+            }
+          )
+        rescue ActiveRecord::RecordNotUnique => e
+          puts e
+        else
+          puts "\t\t- Collection: #{activity.collection_id} \t#{activity.minutes} minutes\n"
         end
+      end
     end
 
-    def self.total_contiguous_seconds(times_and_names, tolerance=90.minutes)
-        total_seconds = 0
-        from_time = nil
+    latest_ahoy_activity_summary = AhoyActivitySummary.order(date: :desc).first
+    keep_after_date = [AhoyActivitySummary::KEEP_AFTER.ago.to_date, latest_ahoy_activity_summary.date.yesterday].min
+
+    exclude_actions = AhoyActivitySummary::WEEKLY_TRIAL_COHORT_TARGET_ACTIONS +
+                      AhoyActivitySummary::WEEKLY_TRANSCRIBER_COHORT_TARGET_ACTIONS
+
+    Ahoy::Event.where('date < ?', keep_after_date).where.not(name: exclude_actions).destroy_all
```

Comment:
Can you let me know if this looks right to you @benwbrum ? 

In ticket it was mentioned we delete Ahoy::Events and Visits, but it seems there is no 'action'  field for Visits. Do I delete only based on `started_at` for Visits then?

Reply:
This looks correct to me.

On events versus visits, one `Visit` may have many `Ahoy::Event` child records.  Generally speaking, a Visit corresponds to a user session, while an Event corresponds to a click or HTTP request.

You can see this in action if you go to the Admin dashboard and click on the _Users_ tab, then click _Visits_ for a user, and drill into a few Visits.

I'm trying to think about the ways we might clean up Visit records.  Since one is created for each session--including anonymous sessions or bots/spiders--we do want to clean these as well.  We certainly want to delete Visit records that have no child Ahoy::Event records after the Event clean-up.

Any ideas on how to do this in a performant manner?